doc/man/anomaly.1.in
.TH anomaly 1 2016-09-04 "${PACKAGE_STRING}" "User Manuals"

.SH NAME
anomaly \- anomalous data detection

.SH SYNOPSIS
.B anomaly [-h|--help] [-v|--version] [-d|--details]
        [-t|--threshold] [--min N] [--max N]
        [-s|--stddev] [-n|--sample N] [-c|--coefficient N]
        [-q|--quiet]
        [-e|--execute PROGRAM]
        [-p|--pid PID]

.SH DESCRIPTION
Anomaly can detect anomalous data in a numeric stream.  In order to do this,
anomaly needs to see a stream of numeric data, and apply one of its detection
methods.  If an anomaly is detected, a response is made, chosen from one or more
built in methods.

.SH NUMERIC STREAM
Anomaly works best in a pipe, and will read only numeric data from its input.
As a simple example, suppose you wish to monitor load average and look for
unusual spikes.  The load average can be obtained from the 'uptime' command:

.RS
$ uptime
.br
11:40  up 15 days,  4:04, 6 users, load averages: 0.38 0.32 0.32
.RE

We can extract the 5-minute load (the second of the three numbers) using this:

.RS
$ uptime | cut -f 13 -d ' '
.br
0.29
.RE

That number can be extracted once a minute, using this:

.RS
$ while [ 1 ]; do uptime | cut -f 13 -d ' '; sleep 60; done
.br
0.29
.br
0.26
.br
0.19
.RE

That is the kind of data stream that anomaly monitors.  White space (spaces,
tabs, newlines) between the numbers are ignored, so we can simulate the above
stream like this:

.RS
$ echo 0.29 0.26 0.19
.RE

This is a convenient way to demonstrate anomaly, shown below.

.SH DETECTION - THRESHOLD
The simplest detection method is threshold, which compares the data to an
absolute value.  This method can use a minimum and a maximum value for
comparison.  These alternatives are all valid, and make use of --min, --max or
both:

.RS
anomaly --threshold --min 1.22 --max 9.75
.br
anomaly --threshold --min 1.22
.br
anomaly --threshold --max 9.75
.RE

In the following example, the values '1' and '10' would be detected as
anomalies:

.RS
$ echo 2 1 3 6 10 5 | anomaly --threshold --min 1.5 --max 8
.br
Anomalous data detected.  The value 1 is below the minimum of 1.5.
.br
Anomalous data detected.  The value 10 is above the maximum of 8.
.RE

.SH DETECTION - STANDARD DEVIATION
Standard deviation measures differences from the mean value of a sample of data,
and is useful for detecting extraordinary values.  The sample size can be chosen
such that there is enough data to determine a good mean value, but defaults to
10.  The limited sample size means that a rolling window of data is used, and
therefore the mean and standard deviation is updated for the current window.
This makes the monitoring somewhat adaptive.  Here is an example:

.RS
anomaly --stddev --sample 20
.RE

This uses a sample size of the 20 most recent values, and will detect any values
that are +/- 1 standard deviation from the mean.  An example:

.RS
$ echo 1 2 3 4 5 6 | anomaly --stddev --sample 5
.br
Anomalous data detected.  The value 6 is more than 1 sigma(s)
above the mean value 3, with a sample size of 5.
.RE

With a sample size of 5, comparisons begin only after the 6th value is seen.  In
the example, the mean value of [1 2 3 4 5] is 3, and the standard deviation is
1.58.  This means that the 6th value is considered an anomaly if it is outside
the range (3 +/- 1.58), which is from 1.42 to 4.58.

To make this less sensitive, a coefficient is introduced, which defaults to 1.0
(as above) but can be overridden:

.RS
$ echo 1 2 3 4 5 6 | anomaly --stddev --sample 5 --coefficient 1.9
.br
$
.RE

In this example, the 6th value is not considered an anomaly because it is within
the range (3 +/- (1.9 * 1.58)), which is between -0.002 and 6.002.

.SH RESPONSE - MESSAGE
The message response is the default, and consists of a single line of printed
text.  It is a description of why the data value is considered an anomaly.  Here
is an example:

.RS
$ echo 1 2 3 | anomaly --threshold --max 2.5
.br
Anomalous data detected.  The value 3 is above the maximum of 2.5.
.RE

The message can be suppressed, but another response must be specified, so that
there is some kind of response:

.RS
$ echo 1 2 3 | anomaly --threshold --max 2.5 --quiet ...
.RE

.SH RESPONSE - EXECUTE
Anomaly can execute a program in response to detection.  Here an example uses
the 'date' command, but any program can be used:

.RS
$ echo 1 2 3 | anomaly --threshold --max 2.5 --quiet --execute '/bin/date +%s'
.br
1361727327
.RE

.SH RESPONSE - SIGNAL
Anomaly can send a USR1 signal to a program in response to detection:

.RS
$ echo 1 2 3 | anomaly --threshold --max 2.5 --quiet --pid 12345
.RE

This sends the USR1 signal to the process with PID 12345.  The receiving program
would need to respond accordingly.

.SH "CREDITS & COPYRIGHTS"
Copyright (C) 2013 - 2017 Göteborg Bit Factory.

Anomaly is distributed under the MIT license. See
http://www.opensource.org/licenses/mit-license.php for more information.

.SH SEE ALSO

For more information, see:

.TP
The official site at
<http://tasktools.org/projects/anomaly.html>

.TP
You can contact the project by writing an email to
<support@tasktools.org>

.SH REPORTING BUGS
.TP
Bugs in anomaly may be reported to <support@tasktools.org>



NEWS

New Features in anomaly 1.2.0

  - 

  Please refer to the ChangeLog file for full details.

---

Anomaly has been built and tested on the following configurations:

  * Arch
  * Cygwin
  * Debian
  * Fedora
  * FreeBSD
  * Ubuntu
  * OS X

---

While Anomaly has undergone testing, bugs are sure to remain.  If you find
a bug, please enter a new issue at:

  https://bug.tasktools.org

Or you can also report the issue in the forums at:

  https://answers.tasktools.org

Or just send a message to:

  support@taskwarrior.org

Thank you.


src/anomaly.cpp
////////////////////////////////////////////////////////////////////////////////
//
// Copyright 2013 - 2017, Göteborg Bit Factory.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//
// http://www.opensource.org/licenses/mit-license.php
//
////////////////////////////////////////////////////////////////////////////////

#include <cmake.h>
#include <iostream>
#include <stdlib.h>
#include <string.h>
#include <Detector.h>
#include <commit.h>

////////////////////////////////////////////////////////////////////////////////
static int usage ()
{
  std::cout << "\n"
               "Usage: anomaly [options]\n"
               "\n"
               "Options:\n"
               "  -h|--help               Show this message\n"
               "  -v|--version            Show anomaly version & copyright\n"
               "  -d|--details            Show internal details\n"
               "\n"
               "Detection: Threshold\n"
               "  -t|--threshold          Threshold algorithm\n"
               "     --min N              Lower threshold\n"
               "     --max N              Upper threshold\n"
               "\n"
               "Detection: Standard Deviation\n"
               "  -s|--stddev             Standard Deviation algorithm (default)\n"
               "  -n|--sample N           Sample size (default: 10)\n"
               "  -c|--coefficient N      Coefficient for --stddev (default: 1.0)\n"
               "\n"
               "Responses:\n"
               "  -q|--quiet              Suppresses anomaly description output\n"
               "  -e|--execute PROGRAM    Runs program\n"
               "  -p|--pid PID            Sends USR1 signal to pid\n"
               "\n";

  return 0;
}

////////////////////////////////////////////////////////////////////////////////
static int version ()
{
  std::cout << "\n"
            << PACKAGE_STRING
#if defined (HAVE_COMMIT)
            << "."
            << COMMIT
#endif
            << " built for "
#if defined (DARWIN)
            << "Darwin"
#elif defined (SOLARIS)
            << "Solaris"
#elif defined (CYGWIN)
            << "Cygwin"
#elif defined (HAIKU)
            << "Haiku"
#elif defined (OPENBSD)
            << "OpenBSD"
#elif defined (FREEBSD)
            << "FreeBSD"
#elif defined (NETBSD)
            << "NetBSD"
#elif defined (LINUX)
            << "Linux"
#elif defined (KFREEBSD)
            << "GNU/kFreeBSD"
#elif defined (GNUHURD)
            << "GNU/Hurd"
#else
            << "unknown"
#endif
            << "\n"
               "Copyright (C) 2013 - 2017 Göteborg Bit Factory\n"
               "\n"
               "Anomaly may be copied only under the terms of the MIT "
               "license, which may be found in the source kit.\n"
               "\n"
               "Documentation for anomaly can be found using 'man anomaly' "
               "or at http://tasktools.org/projects/anomaly.html\n"
               "\n";

  return 0;
}

////////////////////////////////////////////////////////////////////////////////
int main (int argc, char** argv)
{
  try
  {
    Detector detector;

    // Process command line arguments.
    for (int i = 1; i < argc; ++i)
    {
      if (!strcmp (argv[i], "-h") || !strcmp (argv[i], "--help"))
        return usage ();

      else if (!strcmp (argv[i], "-v") || !strcmp (argv[i], "--version"))
        return version ();

      else if (!strcmp (argv[i], "-d") || !strcmp (argv[i], "--details"))
        detector.debug ();

      else if (!strcmp (argv[i], "-t") || !strcmp (argv[i], "--threshold"))
        detector.algorithm ("threshold");

      else if (!strcmp (argv[i], "--max"))
        detector.max (strtod (argv[++i], nullptr));

      else if (!strcmp (argv[i], "--min"))
        detector.min (strtod (argv[++i], nullptr));

      else if (!strcmp (argv[i], "-s") || !strcmp (argv[i], "--stddev"))
        detector.algorithm ("stddev");

      else if (!strcmp (argv[i], "-n") || !strcmp (argv[i], "--sample"))
        detector.sample (strtol (argv[++i], nullptr, 10));

      else if (!strcmp (argv[i], "-c") || !strcmp (argv[i], "--coefficient"))
        detector.coefficient (strtod (argv[++i], nullptr));

      else if (!strcmp (argv[i], "-q") || !strcmp (argv[i], "--quiet"))
        detector.quiet ();

      else if (!strcmp (argv[i], "-p") || !strcmp (argv[i], "--pid"))
        detector.pid (strtol (argv[++i], nullptr, 10));

      else if (!strcmp (argv[i], "-e") || !strcmp (argv[i], "--execute"))
        detector.execute (argv[++i]);

      else
        throw std::string ("Unrecognized argument '") + argv[i] + "'.";
    }

    // Dispatch to selected algorithm.
    detector.run ();
  }

  catch (const std::string& error)
  {
    std::cout << error << "\n";
    return -1;
  }

  catch (...)
  {
    std::cout << "Unknown error\n";
    return -2;
  }

  return 0;
}

////////////////////////////////////////////////////////////////////////////////

